src/GIL/GIL.jl

```diff
@@ -84,13 +145,29 @@ The macro equivalent of [`unlock`](@ref).
 """
 macro unlock(expr)
     quote
-        state = C.PyEval_SaveThread()
+        _locked = Base.islocked(_jl_gil_lock)
+        _locked && Base.unlock(_jl_gil_lock)
         try
-            $(esc(expr))
+            state = C.PyEval_SaveThread()
+            try
+                $(esc(expr))
+            finally
+                C.PyEval_RestoreThread(state)
+            end
         finally
-            C.PyEval_RestoreThread(state)
+            _locked && Base.lock(_jl_gil_lock)
         end
     end
 end
 
+#=
+# Disable this for now since holding this lock will disable finalizers
+# If the main thread already has the GIL, we should lock _jl_gil_lock.
+function __init__()
+    if hasgil()
+        Base.lock(_jl_gil_lock)
+    end
+end
```

Comment:
Locking on `__init__` is probably a bad idea even though we are technically holding the GIL since this will disable finalizers.

Rather we should consider invoking `lock` while executing other calls.

Reply:
Disabling this apparently allowed the python tests to succeed.

Reply:
`OncePerThread` is to be introduced in Julia 1.12 which has yet to be released at the time of this writing. Is there a lesser version of this that could be implemented prior to Julia 1.12, preferably to Julia 1.10 LTS?

Reply:
Yes, Base.Condition, since this is locked to thread-local state